semantic-conventions/src/dynatrace/semconv/templating/markdown.py

```diff
@@ -0,0 +1,404 @@
+#   Copyright 2020 Dynatrace LLC
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import glob
+import io
+import os
+import re
+import sys
+import typing
+from pathlib import PurePath
+
+from dynatrace.semconv.model.constraints import AnyOf, Include
+from dynatrace.semconv.model.semantic_attribute import (
+    SemanticAttribute,
+    EnumAttributeType,
+    Required,
+    EnumMember,
+)
+from dynatrace.semconv.model.semantic_convention import (
+    SemanticConventionSet,
+    SemanticConvention,
+)
+from dynatrace.semconv.model.utils import ID_RE
+
+
+class RenderContext:
+    is_full: bool
+    is_remove_constraint: bool
+    group_key: str
+    break_counter: int
+    enums: list
+    notes: list
+    note_index: int
+    current_md: str
+    current_semconv: SemanticConvention
+
+    def __init__(self, break_count):
+        self.is_full = False
+        self.is_remove_constraint = False
+        self.group_key = ""
+        self.break_count = break_count
+        self.enums = []
+        self.notes = []
+        self.note_index = 1
+        self.current_md = ""
+        self.current_semconv = None
+
+    def clear_table_generation(self):
+        self.notes = []
+        self.enums = []
+        self.note_index = 1
+
+    def add_note(self, msg: str):
+        self.note_index += 1
+        self.notes.append(msg)
+
+    def add_enum(self, attr: SemanticAttribute):
+        self.enums.append(attr)
+
+
+class MarkdownRenderer:
```

Comment:
Since we are using a markdown linter in the spec repository, I preferred building the string via code to have better control over newlines and spaces. I agree that this could be done using Jinja templates. However, I believe it will just push the complexity from code to the template.

Reply:
Indeed we had that discussion internally, but usually the more complicated the string building, the less it profits from a Jinja template, which is fine for building large but more or less simple strings. Just imagine how many nesting levels or macros you would have.